src/OpenTelemetry/Logs/Pool/LogRecordSharedPool.cs

```diff
@@ -0,0 +1,172 @@
+        }
+
+        internal int Count => (int)(Volatile.Read(ref this.returnIndex) - Volatile.Read(ref this.rentIndex));
+
+        public LogRecord Rent()
+        {
+            while (true)
+            {
+                var rentSnapshot = Volatile.Read(ref this.rentIndex);
+                var returnSnapshot = Volatile.Read(ref this.returnIndex);
+
+                if (rentSnapshot >= returnSnapshot)
+                {
+                    break; // buffer is empty
+                }
+
+                if (Interlocked.CompareExchange(ref this.rentIndex, rentSnapshot + 1, rentSnapshot) == rentSnapshot)
+                {
+                    var logRecord = Interlocked.Exchange(ref this.pool[rentSnapshot % this.Capacity], null);
+                    if (logRecord == null && !this.TryRentCoreRare(rentSnapshot, out logRecord))
+                    {
+                        continue;
+                    }
+
+                    logRecord.ResetReferenceCount();
+                    return logRecord;
+                }
+            }
+
+            var newLogRecord = new LogRecord();
+            newLogRecord.ResetReferenceCount();
+            return newLogRecord;
+        }
+
+        public void Return(LogRecord logRecord)
+        {
+            if (logRecord.RemoveReference() != 0)
+            {
+                return;
+            }
+
+            Clear(logRecord);
+
+            while (true)
+            {
+                var rentSnapshot = Volatile.Read(ref this.rentIndex);
+                var returnSnapshot = Volatile.Read(ref this.returnIndex);
+
+                if (returnSnapshot - rentSnapshot >= this.Capacity)
+                {
+                    return; // buffer is full
+                }
+
+                if (Interlocked.CompareExchange(ref this.returnIndex, returnSnapshot + 1, returnSnapshot) == returnSnapshot)
+                {
+                    // If many threads are hammering rent/return it is possible
+                    // for two threads to write to the same index. In that case
+                    // only one of the logRecords will make it back into the
+                    // pool. Anything lost in the race will collected by the GC
+                    // and the pool will issue new instances as needed
+                    this.pool[returnSnapshot % this.Capacity] = logRecord;
+                    return;
+                }
+            }
+        }
+
+        internal static void Clear(LogRecord logRecord)
+        {
+            var attributeStorage = logRecord.AttributeStorage;
+            if (attributeStorage != null)
+            {
+                if (attributeStorage.Count > DefaultMaxNumberOfAttributes)
+                {
+                    // Don't allow the pool to grow unconstained.
+                    logRecord.AttributeStorage = null;
+                }
+                else
+                {
+                    /* List<T>.Clear sets the size to 0 but it maintains the
```

Comment:
nit: rename "size" to "Count".

Reply:
I updated the comment so hopefully it is more clear now.